### ISSUE TITLE
Rest API: Get all folders curl command throwing authorization error

### ISSUE THREAD

**Issue by dineshr93** [fossology/fossology#1706]
### Description

We have deployed our fossology http://fossology.server02.com:8081/repo
we have enabled ldap feature.

I just followed how to list all folders in https://www.fossology.org/get-started/basic-rest-api-calls/

I formed the curl command to list all folders

> curl -k -s -S -X GET "http://fossology.server02.com:8081/repo/api/v1/folders" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJZUzI1NiJ9.eyJpc3MiOiJmb3Nzb2xvZ3kuZWJncm91cC5lbGVrdHJvYml0LmNvbSIsImF1ZCI6ImZvc3NvbG9neS5lYmdyb3VwLmVsZWt0cm9iaXQuY29tIiwiZXhwIjoxNTkwNjIzOTk5LCJuYmYiOjE1ODc5NDU2MDAsImp0aSI6Ik1pNDAiLCJzY29wZSI6IndyaXRlIn0.s_vNuYKamN2ntF33OWQx6Tgn3UXR44K-o0FGVJDRXYE"

We get these ###error

> <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
> <html><head>
> <title>401 Unauthorized</title>
> </head><body>
> <h1>Unauthorized</h1>
> <p>This server could not verify that you
> are authorized to access the document
> requested.  Either you supplied the wrong
> credentials (e.g., bad password), or your
> browser doesn't understand how to supply
> the credentials required.</p>
> <hr>
> <address>Apache/2.4.25 (Debian) Server at fossology.server02.com Port 8081</address>
> </body></html>
> 

Am i making anything wrong?

**Comment by GMishx** [comment]
Hello @dineshr93 ,

Have you tried the LDAP with REST API https://github.com/fossology/fossology/wiki/External-authentication-configuration#ldap-with-rest-api ?

Because it looks like your apache is blocking the request instead of FOSSology.

**Comment by dineshr93** [comment]
Yes 

> Listen 8081

instead of 8880 as mentioned in [ldap rest api](https://github.com/fossology/fossology/wiki/External-authentication-configuration#ldap-with-rest-api )
as my url is http://fossology.server02.com:**8081**/repo

Should I change the port?

@NicolasToussaint

**Comment by NicolasToussaint** [comment]
Hi @dineshr93 ,
The idea of the example Apache conf is that:
- web site is served on  http://fossology.server02.com:<port>/repo
- API is serrved on  http://fossology.server02.com:<port>/api
- -> on the same public port but two distinct URLs

So the REST URL you must use is:
- "fossology.server02.com:8081/**api**/v1/folders"
- and **not** "fossology.server02.com:8081/**repo/api**/v1/folders"

**Comment by dineshr93** [comment]
@NicolasToussaint  I have enabled 
Working url
for web: http://fossology.ebgroup.elektrobit.com:8081/repo/
http://fossology.ebgroup.elektrobit.com:8081/repo/api(browser responds with json)

tried:
web: http://fossology.server02.com:/repo (throws internal server error)
api:http://fossology.server02.com:/api (throws internal server error)


My exact fossology.conf


Listen 8081
```
<VirtualHost *:80>
        ServerName http://fossology.server02.com
        Alias /repo /usr/local/share/fossology/www/ui

        <Location "/api">
                ProxyPass http://fossology.server02.com:8081/repo/api
                ProxyPassReverse http://fossology.server02.com:8081/repo/api
        </Location>

        <Location "/repo">
                Options Indexes FollowSymLinks MultiViews ExecCGI
                AllowOverride None
                Order allow,deny
                allow from all

                AuthLDAPBindDN "username"
                AuthLDAPBindPassword "password"

                AuthType Basic
                AuthName "Required Authentication"
                AuthBasicProvider ldap
                AuthLDAPDereferenceAliases never
                LDAPReferrals off
                AuthLDAPUrl "ldap://ldap_url" NONE
                AuthLDAPGroupAttributeIsDN on
                AuthLDAPGroupAttribute member
                AuthLDAPMaxSubGroupDepth 1
                Require valid-user
                AuthUserFile /dev/null
</Location>
</VirtualHost>

<VirtualHost 127.0.1.1:8081>
        ServerName http://fossology.server02.com
        Alias /repo /usr/local/share/fossology/www/ui
        <Location "/repo">
                Options Indexes FollowSymLinks MultiViews ExecCGI
                AllowOverride None
                Order allow,deny
                allow from all
        </Location>
</VirtualHost>
```

It is enabled as it is present in etc/apache2/sites-enabled inside docker container

I am not sure what I am doing wrong

**Comment by NicolasToussaint** [comment]
> for api: http://fossology.ebgroup.elektrobit.com:8081/repo/api (browser responds with json)

Every URL under `http://<fqdn.com>/repo` will be subject to LDAP/Apache authentication.

To be able to access the REST API **without** LDAP authentication, you need to use `http://<fqdn.com>/api`, and **not** `http://<fqdn.com>/repo/api`

**Comment by dineshr93** [comment]
> To be able to access the REST API **without** LDAP authentication, you need to use `http://<fqdn.com>/api`, and **not** `http://<fqdn.com>/repo/api`


@NicolasToussaint Understood. I have done some mistake. not sure what it is. 

So far I have done done only [ldap rest api](https://github.com/fossology/fossology/wiki/External-authentication-configuration#ldap-with-rest-api )
But 
> Working url
> for web: http://fossology.server02.com:8081/repo/ (browser responds with fossy front page & LDAP login is working fine)
> for api: http://fossology.server02.com:8081/repo/api (browser responds with json)

> web: http://fossology.server02.com:/repo (throws internal server error)
> api:http://fossology.server02.com:/api (throws internal server error)



I haven't done [shibboleth](https://github.com/fossology/fossology/wiki/External-authentication-configuration#shibboleth) procedure. Should I create it and a2ensite it along fossology.conf?

**Comment by dineshr93** [comment]
@NicolasToussaint For those who installs via docker It seems whatever port you are listening in fossology.conf must be exposed in [docker-compose.yml#L52](https://github.com/fossology/fossology/blob/795b19047a857ed8e622208bd74d689c8eefe4c2/docker-compose.yml#L52)

So if in fossology.conf `Listen 8880` is mentioned

then we must add `- 8880:8880 `  below line [L52](https://github.com/fossology/fossology/blob/795b19047a857ed8e622208bd74d689c8eefe4c2/docker-compose.yml#L52). 
port format [- (port that is accessed outside):(port mentioned in Listen of fossology.conf)]

By doing so now I am able to access fossology page that is using normal login instead of ldap login. in http://fossology.server02.com:8880/repo/api

**Comment by NicolasToussaint** [comment]
Indeed, you can also use a separate port to access the API url without LDAP authentication.
If you do that, you can even remove the redirection in the Apache configuration file.

That was not my setup because I didn't want to use a distinct port, but if that works for you, perfect !
